### PR TITLE
feature: Kinesis Firehose Dynamic Partitioning

### PR DESCRIPTION
**Pull request by rjjaegeraws**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Kinesis Firehose Dynamic Partitioning

This pull request depends on [PR#853](https://github.com/aws-samples/aws-secure-environment-accelerator/pull/853).

Currently, all CloudWatch logs are sent to Kinesis Firehose in the log-archive account where it is put into the central log bucket with a prefix similar to the following:
```
CloudWatchLogs/2021/11/09/05/ASEA-Firehose-Delivery-Stream-1-2021-11-09-05-00-50-a023088c-2ed6-4c6b-89e6-aa649a5bd690
``` 

In that gzip'd text file contains an array of records from CloudWatch Logs. Here's an example of one entry from a sample file:
```
{
  "messageType": "DATA_MESSAGE",
  "owner": "370xxxxxx539",
  "logGroup": "/ASEA/rql/Central-vpc-0b5afc0e6b33133bb",
  "logStream": "vpc-0b5afc0e6b33133bb_20211115T1600Z_i-042ddf3855a3d4ed1",
  "subscriptionFilters": [
    "CentralLogging/ASEA/rql/Central-vpc-0b5afc0e6b33133bb"
  ],
  "logEvents": [
    {
      "id": "36506141640138258229402202434304322746355480367892201472",
      "timestamp": 1636992007000,
      "message": "{\"version\":\"1.100000\",\"account_id\":\"370xxxxxx539\",\"region\":\"ca-central-1\",\"vpc_id\":\"vpc-0b5afc0e6b33133bb\",\"query_timestamp\":\"2021-11-15T16:00:07Z\",\"query_name\":\"asea-logarchive-phase0-cacentral1-lgh04fj6ulma.s3.ca-central-1.amazonaws.com.\",\"query_type\":\"A\",\"query_class\":\"IN\",\"rcode\":\"NOERROR\",\"answers\":[{\"Rdata\":\"s3-r-w.ca-central-1.amazonaws.com.\",\"Type\":\"CNAME\",\"Class\":\"IN\"},{\"Rdata\":\"52.95.146.88\",\"Type\":\"A\",\"Class\":\"IN\"}],\"srcaddr\":\"10.1.101.19\",\"srcport\":\"19907\",\"transport\":\"UDP\",\"srcids\":{\"instance\":\"i-042ddf3855a3d4ed1\"}}"
    }
  ]
}
```

PR#853 implements OpenSearch SIEM that ingests data from the log-archive bucket. The 'processing lambda' makes decisions based on the s3prefix key. For example, it looks for 'vpcflowlogs', 'guardduty', etc... in the S3 prefix/key. If the file contains that key, the processing lambda knows how to parse the file and load it into the correct OpenSearch index.

In the current state, it would not be possible for the current OpenSearch SIEM implementation to ingest 'rql' (route53 query) records. This is because these entries are bundled into the generic s3 key; the lambda processor ignores this.

This pull request will recreate the Kinesis Firehose that exists in the log-archive account to enable Dynamic Partitioning. There is a custom lambda as part of this pull requests that executes on records in Kinesis Firehose. It inspects each record entry looking for the log group name. Using a new custom mapping configuration, the lambda will return each record's new S3prefix (or keep the old format as default). In the example above and with the new configuration, the above rql entry would be extracted and would have a different s3 prefix:
```
CloudWatchLogs/rql/2021/11/09/05/ASEA-Firehose-Delivery-Stream-1-2021-11-09-05-00-50-a023088c-2ed6-4c6b-89e6-aa649a5bd690
``` 

If the OpenSearch SIEM processing lambda has the proper entry in its INI file (ex: watching for 'rql'), then this record type will be loaded into OpenSearch.

The new configuration (**dynamic-s3-log-partitioning**) looks like the following:
```
 "central-log-services": {
      "account": "log-archive",
      "region": "ca-central-1",
      "s3-retention": 730,
      "cwl-glbl-exclusions": ["/ASEA/flowlogs/*", "/ASEA/CloudTrail*"],
      "cwl-exclusions": [],
      "ssm-to-s3": true,
      "ssm-to-cwl": true,
      "sns-excl-regions": [],
      "sns-subscription-emails": {
        "High": ["notify-high@example.com"],
        "Medium": ["notify-medium@example.com"],
        "Low": ["notify-low@example.com"]
      },
      "dynamic-s3-log-partitioning": [
        {
          "logGroupPattern": "MAD",
          "s3Prefix": "managed-ad"
        },
        {
          "logGroupPattern": "rql",
          "s3Prefix": "rql"
        },
        {
          "logGroupPattern": "SecurityHub",
          "s3Prefix": "security-hub"
        }        
      ]
    },
```

As an example, this entry:
```
{
    "logGroupPattern": "MAD",
    "s3Prefix": "managed-ad"
},
```
will have the lambda check for 'MAD' in the log group name, if it is found, then that record will have a prefix of 'managed-ad'.

### CODE REVIEW

**Comment by Brian969** [comment]
Question - this `"logGroupPattern": "MAD"` is designed for log groups in the format `/${ACCELERATOR_PREFIX_ND}/{log-group-type}/*` - should this be opened up to enable this new feature to be used by *any* customer log group prefix?  i.e. change `"logGroupPattern": "MAD"` to `"logGroupPattern": "/ASEA/MAD"` and adjust code respectively?

File: reference-artifacts/SAMPLE_CONFIGS/config.example.json [hunk 1]
```diff
@@ -93,7 +93,21 @@
...
+          "s3Prefix": "managed-ad"
+        },
+        {
+          "logGroupPattern": "rql",
+          "s3Prefix": "rql"
+        },
+        {
+          "logGroupPattern": "SecurityHub",
+          "s3Prefix": "security-hub"
+        }
```
Comment (Brian969) [inline reference-artifacts/SAMPLE_CONFIGS/config.example.json:109]: Can we add these 3 additional log types:
,
        {
          "logGroupPattern": "Nfw",
          "s3Prefix": "nfw"
        },
        {
          "logGroupPattern": "rsyslog",
          "s3Prefix": "rsyslog"
        },
        {
          "logGroupPattern": "SSM",
          "s3Prefix": "ssm"
        }

**Comment by rjjaegeraws** [comment]
> /${ACCELERATOR_PREFIX_ND}/

Yes, this is probably better. I've noticed that for the SecurityHub it could also pickup supporting Lambda CWLs, which we would want out of there.

Something like this would be better:
```
 {
          "logGroupPattern": "/${ACCELERATOR_PREFIX_ND}/MAD",
          "s3Prefix": "managed-ad"
        },
        {
          "logGroupPattern": "/${ACCELERATOR_PREFIX_ND}/rql",
          "s3Prefix": "rql"
        },
        {
          "logGroupPattern": "/${ACCELERATOR_PREFIX_ND}/SecurityHub",
          "s3Prefix": "security-hub"
        },
```

**Comment by Brian969** [comment]
> > /${ACCELERATOR_PREFIX_ND}/
> 
> Yes, this is probably better. I've noticed that for the SecurityHub it could also pickup supporting Lambda CWLs, which we would want out of there.
> 
> Something like this would be better:
> 
> ```
>  {
>           "logGroupPattern": "/${ACCELERATOR_PREFIX_ND}/MAD",
>           "s3Prefix": "managed-ad"
>         },
>         {
>           "logGroupPattern": "/${ACCELERATOR_PREFIX_ND}/rql",
>           "s3Prefix": "rql"
>         },
>         {
>           "logGroupPattern": "/${ACCELERATOR_PREFIX_ND}/SecurityHub",
>           "s3Prefix": "security-hub"
>         },
> ```

Agreed!

**Comment by Brian969** [comment]
NOTE: This PR encompasses PR#853.  Only PR#861 (this) needs to be merged.

**Review by hickeydh-aws — APPROVED** [review]

**Pull request merged**